allenact/base_abstractions/sensor.py

```diff
@@ -192,38 +231,63 @@ def _get_observation_space(self) -> gym.spaces.Box:
         only if the expert failed to generate a true expert action. The
         value `num actions in task` should be in `config["nactions"]`
         """
-        return su.flatten_space(self.unflattened_observation_space)
+        return self.flagged_space(self.action_space, use_dict_as_groups=self.use_groups)
 
     @lazy_property
-    def _zeroed_observation(self):
-        return np.zeros_like(self.observation_space.sample())
+    def _zeroed_observation(self) -> Union[OrderedDict, Tuple]:
+        # AllenAct-style flattened space (to easily generate an all-zeroes action as an array)
+        flat_space = su.flatten_space(self.observation_space)
+        # torch point to correctly unflatten `Discrete` for zeroed output
+        flat_zeroed = su.torch_point(flat_space, np.zeros_like(flat_space.sample()))
+        # unflatten zeroed output and convert to numpy
+        return su.numpy_point(
+            self.observation_space, su.unflatten(self.observation_space, flat_zeroed)
+        )
+
+    def flatten_output(self, unflattened):
+        return (
+            su.flatten(
+                self.observation_space,
+                su.torch_point(self.observation_space, unflattened),
+            )
+            .cpu()
+            .numpy()
+        )
 
     def get_observation(
         self, env: EnvType, task: SubTaskType, *args: Any, **kwargs: Any
-    ) -> Any:
+    ) -> Union[OrderedDict, Tuple]:
         # If the task is completed, we needn't (perhaps can't) find the expert
         # action from the (current) terminal state.
         if task.is_done():
-            return self._zeroed_observation
-
-        action, expert_was_successful = task.query_expert(**self.expert_args)
+            return self.flatten_output(self._zeroed_observation)
 
-        if isinstance(action, int):
-            assert isinstance(self.action_space, gym.spaces.Discrete)
-            unflattened_action = action
-        else:
-            # Assume we receive a gym-flattened numpy action
-            unflattened_action = gyms.unflatten(self.action_space, action)
+        actions = OrderedDict()
+        for group_name in self.group_spaces:
+            action, expert_was_successful = task.query_expert(
+                **self.expert_args, expert_sensor_action_group_name=group_name
+            )
 
-        unflattened_torch = su.torch_point(
-            self.unflattened_observation_space,
-            (unflattened_action, expert_was_successful),
-        )
+            if isinstance(action, int):
+                assert isinstance(self.group_spaces[group_name], gym.spaces.Discrete)
+                unflattened_action = action
+            else:
+                # Assume we receive a gym-flattened numpy action
+                unflattened_action = gyms.unflatten(
+                    self.group_spaces[group_name], action
+                )
+                # TODO why not enforcing unflattened actions from the task?
```

Comment:
This TODO would seem reasonable to me, is there anything that makes this difficult?